### ISSUE TITLE
yolox train customize dataset the loss_cls: 0.0000, loss_bbox: 0.0000

### ISSUE THREAD

**Issue by EnzoLiang** [open-mmlab/mmdetection#9359]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmdetection/issues) and [Discussions](https://github.com/open-mmlab/mmdetection/discussions) but cannot get the expected help.
- [X] I have read the [FAQ documentation](https://mmdetection.readthedocs.io/en/latest/faq.html) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version (master)](https://github.com/open-mmlab/mmdetection) or [latest version (3.x)](https://github.com/open-mmlab/mmdetection/tree/dev-3.x).

### Task

I have modified the scripts/configs, or I'm working on my own tasks/models/datasets.

### Branch

master branch https://github.com/open-mmlab/mmdetection

### Environment

sys.platform: linux
Python: 3.7.13 (default, Mar 29 2022, 02:18:16) [GCC 7.5.0]
CUDA available: True
GPU 0,1: Tesla P100-PCIE-16GB
CUDA_HOME: /usr/local/cuda-10.0
NVCC: Cuda compilation tools, release 10.0, V10.0.13
GCC: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
PyTorch: 1.7.1
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.6.0 (Git Hash 5ef631a030a6f73131c77892041042805a06064f)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_VULKAN_WRAPPER -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

TorchVision: 0.8.2
OpenCV: 4.6.0
MMCV: 1.7.0
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 10.1
MMDetection: 2.25.3+e71b499

### Reproduces the problem - code sample

_base_ = ['../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py']

img_scale = (640, 640)  # height, width

# model settings
model = dict(
    type='YOLOX',
    input_size=img_scale,
    random_size_range=(15, 25),
    random_size_interval=10,
    backbone=dict(type='CSPDarknet', deepen_factor=1.33, widen_factor=1.25),
    neck=dict(
        type='YOLOXPAFPN',
        in_channels=[320, 640, 1280], out_channels=320, num_csp_blocks=4),
    bbox_head=dict(
        type='YOLOXHead', num_classes=2, in_channels=320, feat_channels=320),
    train_cfg=dict(assigner=dict(type='SimOTAAssigner', center_radius=2.5)),
    # In order to align the source code, the threshold of the val phase is
    # 0.01, and the threshold of the test phase is 0.001.
    test_cfg=dict(score_thr=0.01, nms=dict(type='nms', iou_threshold=0.65)))

# dataset settings
data_root = '/mnt/122.149-sr01/liangyzh/dataset/coco/exam_0409_0418/'
dataset_type = 'CocoDataset'

train_pipeline = [
    dict(type='Mosaic', img_scale=img_scale, pad_val=114.0),
    dict(
        type='RandomAffine',
        scaling_ratio_range=(0.1, 2),
        border=(-img_scale[0] // 2, -img_scale[1] // 2)),
    dict(
        type='MixUp',
        img_scale=img_scale,
        ratio_range=(0.8, 1.6),
        pad_val=114.0),
    dict(type='YOLOXHSVRandomAug'),
    dict(type='RandomFlip', flip_ratio=0.5),
    # According to the official implementation, multi-scale
    # training is not considered here but in the
    # 'mmdet/models/detectors/yolox.py'.
    dict(type='Resize', img_scale=img_scale, keep_ratio=True),
    dict(
        type='Pad',
        pad_to_square=True,
        # If the image is three-channel, the pad value needs
        # to be set separately for each channel.
        pad_val=dict(img=(114.0, 114.0, 114.0))),
    dict(type='FilterAnnotations', min_gt_bbox_wh=(1, 1), keep_empty=False),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
]

train_dataset = dict(
    type='MultiImageMixDataset',
    dataset=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/instances_train2017.json',
        img_prefix=data_root + 'train2017/',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='LoadAnnotations', with_bbox=True)
        ],
        filter_empty_gt=False,
    ),
    pipeline=train_pipeline)

test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=img_scale,
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(
                type='Pad',
                pad_to_square=True,
                pad_val=dict(img=(114.0, 114.0, 114.0))),
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img'])
        ])
]

data = dict(
    samples_per_gpu=4,
    workers_per_gpu=4,
    persistent_workers=True,
    train=train_dataset,
    val=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/instances_val2017.json',
        img_prefix=data_root + 'val2017/',
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/instances_val2017.json',
        img_prefix=data_root + 'val2017/',
        pipeline=test_pipeline))

# optimizer
# default 8 gpu
optimizer = dict(
    type='SGD',
    lr=0.01,
    momentum=0.9,
    weight_decay=5e-4,
    nesterov=True,
    paramwise_cfg=dict(norm_decay_mult=0., bias_decay_mult=0.))
optimizer_config = dict(grad_clip=None)

max_epochs = 400
num_last_epochs = 15
resume_from = None
interval = 1

# learning policy
lr_config = dict(
    _delete_=True,
    policy='YOLOX',
    warmup='exp',
    by_epoch=False,
    warmup_by_epoch=True,
    warmup_ratio=1,
    warmup_iters=5,  # 5 epoch
    num_last_epochs=num_last_epochs,
    min_lr_ratio=0.05)

runner = dict(type='EpochBasedRunner', max_epochs=max_epochs)

custom_hooks = [
    dict(
        type='YOLOXModeSwitchHook',
        num_last_epochs=num_last_epochs,
        priority=48),
    dict(
        type='SyncNormHook',
        num_last_epochs=num_last_epochs,
        interval=interval,
        priority=48),
    dict(
        type='ExpMomentumEMAHook',
        resume_from=resume_from,
        momentum=0.0001,
        priority=49)
]
checkpoint_config = dict(interval=interval)
evaluation = dict(
    save_best='auto',
    # The evaluation interval is 'interval' when running epoch is
    # less than ‘max_epochs - num_last_epochs’.
    # The evaluation interval is 1 when running epoch is greater than
    # or equal to ‘max_epochs - num_last_epochs’.
    interval=interval,
    dynamic_intervals=[(max_epochs - num_last_epochs, 1)],
    metric='bbox')
log_config = dict(interval=50)

# NOTE: `auto_scale_lr` is for automatically scaling LR,
# USER SHOULD NOT CHANGE ITS VALUES.
# base_batch_size = (8 GPUs) x (8 samples per GPU)
auto_scale_lr = dict(base_batch_size=4)

# load_from = "/mnt/122.149-sr01/liangyzh/dataset/mmlab/mmdetection/yolox_x_8x8_300e_coco_20211126_140254-1ef88d67.pth"
work_dir = "/mnt/122.149-sr01/liangyzh/dataset/mmlab/mmdetection/output/20221121"

### Reproduces the problem - command or script

python tools/train.py configs/yolox/yolox_x_1\*8_exam.py --auto-scale-lr --gpu-id 0

### Reproduces the problem - error message

2022-11-21 17:34:04,502 - mmdet - INFO - Epoch [1][50/2381]     lr: 1.764e-07, eta: 7 days, 6:09:22, time: 0.658, data_time: 0.064, memory: 12372, loss_cls: 0.0000, loss_bbox: 0.000$
, loss_obj: 375.9151, loss: 375.9151
2022-11-21 17:34:34,780 - mmdet - INFO - Epoch [1][100/2381]    lr: 7.056e-07, eta: 6 days, 23:10:02, time: 0.606, data_time: 0.006, memory: 14254, loss_cls: 0.0000, loss_bbox: 0.00$
0, loss_obj: 318.3267, loss: 318.3267
2022-11-21 17:35:07,847 - mmdet - INFO - Epoch [1][150/2381]    lr: 1.588e-06, eta: 7 days, 1:44:59, time: 0.661, data_time: 0.006, memory: 14254, loss_cls: 0.0000, loss_bbox: 0.000$
, loss_obj: 259.7934, loss: 259.7934
2022-11-21 17:35:35,278 - mmdet - INFO - Epoch [1][200/2381]    lr: 2.822e-06, eta: 6 days, 19:34:55, time: 0.549, data_time: 0.006, memory: 14254, loss_cls: 0.0000, loss_bbox: 0.000
0, loss_obj: 125.2541, loss: 125.2541
2022-11-21 17:36:01,379 - mmdet - INFO - Epoch [1][250/2381]    lr: 4.410e-06, eta: 6 days, 14:28:18, time: 0.522, data_time: 0.006, memory: 14254, loss_cls: 0.0000, loss_bbox: 0.000
0, loss_obj: 69.2520, loss: 69.2520
2022-11-21 17:36:28,697 - mmdet - INFO - Epoch [1][300/2381]    lr: 6.350e-06, eta: 6 days, 12:08:08, time: 0.546, data_time: 0.006, memory: 14254, loss_cls: 0.0000, loss_bbox: 0.000
0, loss_obj: 44.0446, loss: 44.0446
2022-11-21 17:36:58,185 - mmdet - INFO - Epoch [1][350/2381]    lr: 8.643e-06, eta: 6 days, 12:06:14, time: 0.590, data_time: 0.006, memory: 14254, loss_cls: 0.0000, loss_bbox: 0.000
0, loss_obj: 29.0738, loss: 29.0738

**Comment by EnzoLiang** [comment]
> Maybe you can check that the label format for your custom data is correct

hi @Czm369 , I've checked the label format, the categories id could match with  category_id in annotations .

for example:
"annotations": [
        {
            "area": 20305.0,
            "iscrowd": 0,
            "bbox": [
                885.0,
                519.0,
                131.0,
                155.0
            ],
            "category_id": 1,
            "ignore": 0,
            "image_id": 0,
            "id": 1
        },
    ]

"categories": [
        {
            "supercategory": "none",
            "id": 1,
            "name": "sit"
        },
        {
            "supercategory": "none",
            "id": 2,
            "name": "stand"
        }
    ]

**Comment by EnzoLiang** [comment]
I've solved this problem. I forgot add classes in config.

**Comment by wenyuqing** [comment]
> I've solved this problem. I forgot add classes in config.

Hi，I meet the same problem. I 've already set num_classes to 80 on COCO, but still get zero loss_cls and loss_bbox. Could you elaborate your solution? Thanks.

**Comment by zxywudi** [comment]
> 我已经解决了这个问题。我忘了在配置中添加类。

您好，我在coco.py以及classes.py中修改了类信息，并且在配置文件中将classnum修改成对应的数目，但是依然出现了这个错误，请问您具体是怎么解决的呢

**Comment by cosimo17** [comment]
@EnzoLiang  I meet the same problem. can you share your config? I dont know how to modify the classes.

**Comment by SajjadPSavoji** [comment]
This Issue is a bit overdue, asking in case people are still facing the same problem ...
I have managed to set the `classes` correctly. Despite of that, I'm still getting 0.0 on bbox and classification loss.

Also, on the test step, when it tries to compute the COCO metrics, I get the following error.

" ERROR - ./mmdetection/mmdet/evaluation/metrics/coco_metric.py - compute_metrics - 465 - The testing results of the whole dataset is empty."

Any suggestions how to proceed from here?

**Comment by YuanboZhang666** [comment]
> This Issue is a bit overdue, asking in case people are still facing the same problem ... I have managed to set the `classes` correctly. Despite of that, I'm still getting 0.0 on bbox and classification loss.
> 
> Also, on the test step, when it tries to compute the COCO metrics, I get the following error.
> 
> " ERROR - ./mmdetection/mmdet/evaluation/metrics/coco_metric.py - compute_metrics - 465 - The testing results of the whole dataset is empty."
> 
> Any suggestions how to proceed from here?

Hi , Have you solved your problem? I have the same

**Comment by hanzi326** [comment]
I encountered the same issue. The problem of loss being zero was resolved by turning off the random affine data augmentation. I'm not sure if disabling random affine is the correct way to fix this issue or if there are other solutions available.

**Comment by YuanboZhang666** [comment]
My solution is to run the code in pycharm and set the path of the dataset
in the config file as an absolute path. Problem solved.

HanWu ***@***.***>于2023年12月26日 周二16:16写道：

**Comment by Abhishek0075** [comment]
If still not solved, then add metainfo in the config file

Create a variable :
`metainfo = dict(classes=('your_class_name',), palette=[(20, 220, 60)]) `

Then add metainfo in datasets i.e.

```
train_dataset = dict(
    type=dataset_type,
    data_root=data_root,
    metainfo=metainfo,
    ann_file='your/annotation/path',
    data_prefix=dict(img='your/image/'),
    pipeline=[
        dict(type='LoadImageFromFile', backend_args=backend_args),
        dict(type='LoadAnnotations', with_bbox=True),
        *train_pipeline
    ],
    filter_cfg=dict(filter_empty_gt=False, min_size=32),
    backend_args=backend_args)
```
Then add this train_dataset into train_loader like :-
```
train_dataloader = dict(
    batch_size=4,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    dataset=train_dataset)
```

Similarly add in validation dataloader too
